v2/sourcedb-to-spanner/terraform/samples/sharded-bulk-migration/README.md

```diff
@@ -0,0 +1,435 @@
+
+## Terraform permissions
+
+In order to create the resources in this sample,
+the `Service account`/`User account` being used to run Terraform
+should have the
+required [permissions](https://cloud.google.com/iam/docs/manage-access-service-accounts#multiple-roles-console).
+There are two ways to add permissions -
+
+1. Adding pre-defined roles to the service account running Terraform.
+2. Creating a custom role with the granular permissions and attaching it to the
+   service account running Terraform.
+
+### Using custom role and granular permissions (recommended)
+
+Following permissions are required -
+
+```shell
+- compute.globalAddresses.create
+- compute.globalAddresses.createInternal
+- compute.globalAddresses.delete
+- compute.globalAddresses.deleteInternal
+- compute.globalAddresses.get
+- compute.globalOperations.get
+- compute.networks.addPeering
+- compute.networks.get
+- compute.networks.listPeeringRoutes
+- compute.networks.removePeering
+- compute.networks.use
+- compute.routes.get
+- compute.routes.list
+- compute.subnetworks.get
+- compute.subnetworks.list
+- dataflow.jobs.cancel
+- dataflow.jobs.create
+- dataflow.jobs.updateContents
+- iam.roles.get
+- iam.serviceAccounts.actAs
+- storage.objects.delete
+- storage.objects.create
+- serviceusage.services.use
+- serviceusage.services.enable
+```
+
+**Note**: Add the `roles/viewer` role as well to the service account.
+
+[This](#adding-access-to-terraform-service-account) section in the FAQ
+provides instructions to add these permissions to an existing service account.
+
+### Using pre-defined roles
+
+Following roles are required -
+
+```shell
+roles/dataflow.admin 
+roles/iam.securityAdmin
+roles/iam.serviceAccountUser 
+roles/storage.admin
+roles/viewer
+roles/compute.networkAdmin
+```
+
+[This](#adding-access-to-terraform-service-account) section in the FAQ
+provides instructions to add these roles to an existing service account.
+
+## Dataflow permissions
+
+The Dataflow service account needs to be provided the `roles/storage.objectAdmin` and `roles/spanner.databaseAdmin`
+roles. Also ensure that the workers can access the source shards.
+
+## Assumptions
+
+It makes the following assumptions -
+
+> **_NOTE:_**
+[SMT](https://googlecloudplatform.github.io/spanner-migration-tool/quickstart.html)
+> can be used for converting a MySQL schema to a Spanner compatible schema.
+
+1. Dataflow can establish network connectivity with the MySQL source shards.
+2. Appropriate permissions are added to the service account running Terraform to allow resource creation.
+3. Appropriate permissions are provided to the service account running Dataflow to write to Spanner.
+4. A GCS directory has been provided to write the output records (DLQ/filtered events) to.
+5. A Spanner instance with database containing the data-migration compatible schema is created. This includes a
+   migration_shard_id column being present in the Spanner schema.
```

Comment:
Is this mandatory? What if the user does not want this column at all?

Reply:
This is a mandatory column for sharded migrations AFAIK. They can rename it but not avoid it for sharded flows. And a session file is also mandatory for sharded flows.